### PR TITLE
Umbrella support

### PR DESCRIPTION
**Pull request by bengro**
_This PR rewrites most of the extension's codebase to support umbrella projects_.

**User interface changes**
- The user cannot (doesn't need to) select the project directory manually. Instead, any mix-based projects are automatically found and indexed. Projects are identified by a `mix.exs` file and a neighbouring `test` directory, which is standard for Elixir.

**Conceptual changes**
- The tree structure of test suites and tests is stored in `graphlib`. This means it's very easy to retrieve and update the tree without traversing the Test Adapater objects (TestInfo and TestSuiteInfo).
- I have added tests for large parts of the code base. The parsers remain unchanged and untested.

**Fixed issues**
- https://github.com/adamzapasnik/vscode-elixir-test-explorer/issues/1
- https://github.com/adamzapasnik/vscode-elixir-test-explorer/issues/3

**Notes**
- When files are updated/deleted, instead of reloading only the changed node, the entire test tree is re-loaded. Thanks to the underlying test tree, this is an easy optimisation.
- A few TODOs are left in the codebase, I believe some of them may deserve some issues to be created.
- All verification tasks are green `npm run test-unit`, `npm run test` and `npm run compile`.

I have tested the changes with my umbrella project. I hope it works for you as well as it works for me. Let me know of any feedback or bugs.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -61,28 +55,37 @@
...
     "prettier": "^2.2.1",
+    "ts-mocha": "^8.0.0",
     "typescript": "^4.1.3",
     "vscode-test": "^1.4.1"
   },
   "extensionDependencies": [
     "hbenl.vscode-test-explorer"
   ],
   "dependencies": {
+    "@types/graphlib": "^2.1.7",
```
[the remark this answers is not in the document]
Comment (bengro) [reply package.json:87]: No. This library helps to build the dependency tree of tests and test suites.

File: src/TestTree.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    };
+    this.graph.setNode('ExUnit_suite_root', this.root);
+  }
+
+  public getNode(nodeId: string): Test | TestSuiteInfo {
+    return this.graph.node(nodeId);
+  }
+
+  public addTest(test: Test): Test {
+    const v = test.belongsTo.id || this.root.id;
```
Comment (adamzapasnik) [inline src/TestTree.ts:70]: what does `v` and `w` mean?

**Review by adamzapasnik — COMMENTED** [review]
I'll have one more look at TestTree and tests in the evening.

File: src/ExUnitRunner.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    var results: string[] = [];
+
+    var walk = function (currentDir: string): void {
+      var list = fs.readdirSync(currentDir);
+
+      list.forEach(function (file: string) {
+        file = path.join(currentDir, file);
+        var stat = fs.statSync(file);
+        if (stat && stat.isDirectory() && isNotDepsDir(file)) {
```
[the remark this answers is not in the document]
Comment (bengro) [reply src/ExUnitRunner.ts:87]: Nice, will add.

File: src/TestTree.ts (hunk 2 above)
Comment (bengro) [reply src/TestTree.ts:70]: In graph theory, v and w are standard names for the tail and head of an edge.

https://github.com/dagrejs/graphlib/wiki/API-Reference#hasEdge

File: src/ExUnitRunner.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamzapasnik) [reply src/ExUnitRunner.ts]: one more dir came to my mind, `.elixir_ls`

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamzapasnik) [reply package.json:87]: Yes, but I was asking about the `@types/graphlib`, I thought `@types/...` could be added to dev? Don't have much experience with TS, though, so I'm asking.

File: src/TestTree.ts (hunk 2 above)
Comment (adamzapasnik) [reply src/TestTree.ts:70]: Didn't know that, thanks :)

**Comment by adamzapasnik** [comment]
I just tested it with https://github.com/elixir-ecto/ecto and I think we need to improve two things.

Notice there is a mix project `examples/friends`.

I ran mix deps.get inside of ecto, but tests didn't load, I got this message:
```shell
Failed parsing projects: Command failed: mix test --trace --seed=0 --only="" 
Unchecked dependencies for environment test:
* ecto_sql (Hex package)
  the dependency is not available, run "mix deps.get"
* telemetry (Hex package)
  the dependency is not available, run "mix deps.get"
* decimal (Hex package)
  the dependency is not available, run "mix deps.get"
* jason (Hex package)
  the dependency is not available, run "mix deps.get"
* postgrex (Hex package)
  the dependency is not available, run "mix deps.get"
** (Mix) Can't continue due to errors on dependencies
```
I think we are missing some information here which project is affected. Took me a while to realize there is one more mix.exs.

I'm not sure if I'm a fan of loading all projects within a workspace, but I think it's okay for now. I'm also thinking if there is a simple fix to show loaded tests even if one of the suites failed to load. Currently, the example/friends failed to load and ecto's tests weren't displayed either. If there is no quick win here, let's leave it as it is.

File: src/ExUnitRunner.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  private readonly workspacePath: string;
+  private mixRunner: MixRunner;
+  private mixParser: MixParser;
+  private testTree: TestTree;
+
+  constructor(workspaceName: string, workspacePath: string) {
+    this.mixRunner = new MixRunner();
+    this.mixParser = new MixParser();
+    this.testTree = new TestTree(workspaceName);
+    this.workspaceName = workspaceName;
```
Comment (adamzapasnik) [inline src/ExUnitRunner.ts:30]: this seems unnecessary? It's not used anywhere in this file.

File: src/TestTree.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        return buildSuite(this.root.id);
+      }
+      case 'NODE_ID': {
+        return buildNode(options.nodeId!);
+      }
+    }
+  }
+
+  public print() {
+    console.log('graph:', json.write(this.graph));
```
Comment (adamzapasnik) [inline src/TestTree.ts:233]: do we use it anywhere? :thinking:

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (bengro) [reply package.json:87]: Ahh, yes, you're absolutely right.

File: src/ExUnitRunner.ts (hunk 4 above)
Comment (bengro) [reply src/ExUnitRunner.ts]: Removed!

**Comment by adamzapasnik** [comment]
Awesome! :) Very nice! Could you rebase the branch (conflicts)? I'll have one last look tonight/tomorrow and will try to release it :+1:

**Comment by bengro** [comment]
😎 , thanks! Can't wait to use it 😊

File: src/main.ts [hunk 6]
```diff
@@ -21,8 +22,13 @@ export async function activate(context: vscode.ExtensionContext) {
...
 
     // this will register an ExUnitAdapter for each WorkspaceFolder
+    const exUnitRunner = new ExUnitRunner(workspaceFolder.name, workspaceFolder.uri.fsPath);
     context.subscriptions.push(
-      new TestAdapterRegistrar(testHub, (workspaceFolder) => new ExUnitAdapter(workspaceFolder, log), log)
+      new TestAdapterRegistrar(
+        testHub,
+        (workspaceFolder) => new ExUnitTestAdapter(exUnitRunner, workspaceFolder, log),
+        log
+      )
```
Comment (adamzapasnik) [inline src/main.ts:31]: There is a bug here. Line `25` uses first workspaceFolder to create a runner, where it should use `workspaceFolder` from line `29`. I think you could initialize exUnitRunner inside ExUnitTestsAdapter's constructor.

File: src/main.ts [hunk 7]
```diff
@@ -1,7 +1,8 @@
 import * as vscode from 'vscode';
 import { TestHub, testExplorerExtensionId } from 'vscode-test-adapter-api';
 import { Log, TestAdapterRegistrar } from 'vscode-test-adapter-util';
-import { ExUnitAdapter } from './adapter';
+import { ExUnitTestAdapter } from './ExUnitTestAdapter';
+import { ExUnitRunner } from './ExUnitRunner';
 
 export async function activate(context: vscode.ExtensionContext) {
   const workspaceFolder = (vscode.workspace.workspaceFolders || [])[0];
```
Comment (adamzapasnik) [inline src/main.ts:8]: You can rename this to `logWorkspaceFolder` to make is less confusing.

File: test/unit/TestTree.test.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  describe('test suites data integrity', () => {
+    it('test suites and tests contain correct file, label and id', async () => {
+      tree.import(parseMixOutput(process.cwd(), PATHS.simpleProject, await mix.run(PATHS.simpleProject)));
+
+      const exportedTree = tree.export() as TestSuiteInfo;
+
+      // root
+      expect(exportedTree.id).to.to.equal('ExUnit_suite_root');
+      expect(exportedTree.label).to.to.equal('ExUnit workspaceName');
+      expect(exportedTree.file).to.to.equal(undefined);
```
Comment (adamzapasnik) [inline test/unit/TestTree.test.ts:64]: any reason for double `to`?

File: test/unit/MixRunner.test.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    it('run command without path executes only tests in dir', async () => {
+      const runner = new MixRunner();
+
+      const output = await runner.run(PATHS.simpleProject, '');
+
+      expect(output).to.not.be.empty;
+      expect(output).to.contain('Finished in ');
+
+      expect(output).to.contain('test/nested_dir/nested_test.exs');
+      expect(output).to.contain('(excluded)'); // one of the tests if failing
```
Comment (adamzapasnik) [inline test/unit/MixRunner.test.ts:19]: I think `is` failing?

**Comment by bengro** [comment]
Thorough QA, great stuff, thanks! I will have a look into the issues and suggestions in the next two days.

> Somehow I missed it... But project tests should be nested in a "project" suite. User should be able to execute all single project's tests and it would make it much easier to navigate through them. Now it's hard to tell which app/project they come from.

Would you mind sending me a screenshot for this? For my project it does that I thought ... 🤔

**Comment by adamzapasnik** [comment]
Screenshot from 2021-05-12 18-01-16

What I think would be nice. This way we can just click `app two` and all `app two`'s tests are running. 
```
- exunit umbrella project
- - app two <<<<
- - - app_two_test.exs
- - - - doctest
- - - - greets the world
- - app one <<<<
- - - app_one_test.exs
- - - - doctest
- - - - greets the world
```

**Comment by bengro** [comment]
Gotcha! Will look into it.

**Comment by bengro** [comment]
Screenshot 2021-05-13 at 08 12 26

I have fixed it, I think. Please verify it works as you wished.

**Comment by bengro** [comment]
I just looked into the other issue you spotted where the relative path to `mix` is off. 

I'll look into that tomorrow to wrap everything up. I think it's worth fixing.

**Comment by adamzapasnik** [comment]
Thanks, I'll have a look during the weekend :bow:

**Comment by bengro** [comment]
It's good you spotted these things, this lead to a refactor regarding various file paths.

Let me know if you can spot any more bugs 😬

File: src/MixRunner.ts [hunk 10]
```diff
@@ -29,25 +29,25 @@ export class MixRunner {
   }
 
   // TODO: can these two methods not be merged? I don't understand the difference.
-  public async evaluate(projectDir: string, filePath: string = ''): Promise<string> {
```
Comment (adamzapasnik) [inline src/MixRunner.ts]: could we also add this to `run`? It can't load tests without this.

File: src/ExUnitRunner.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  nodeId: 'string';
+  error?: 'string';
+}
+
+/*
+ This class is responsible for loading and running tests.
+ Under the hood it uses the TestTree and the MixRunner to do so.
+*/
+export class ExUnitRunner {
+  private readonly workspacePath: string;
```
Comment (adamzapasnik) [inline src/ExUnitRunner.ts:20]: isn't used anymore

**Comment by bengro** [comment]
Hmm 🤔 

*Run* works for me ... and it should, because it's already running in the mix directory.

Can you try again please? What's the error you're getting?

**Comment by adamzapasnik** [comment]
Apologies, I had problem with a local environment. It works :)

One more bug to fix and we are ready to go :rocket: 
Add two workspaces and try to run tests by clicking on one of the "root" nodes.
```
Failed to run tests in project ./:
 ** (Mix) Could not find a Mix.Project, please ensure you are running Mix in a directory with a mix.exs file
```
Screenshot from 2021-05-18 19-44-57

**Comment by bengro** [comment]
Hmm, that's a tricky one to solve. I'd rather merge our stuff now and wait for an issue to be raised. Happy to fix it when someone requests it. 

The old version of the plugin doesn't support multiple workspaces, I don't think it's a regression.

**Comment by adamzapasnik** [comment]
> The old version of the plugin doesn't support multiple workspaces, I don't think it's a regression.
It does, though.

But year, fair enough, it's not a must atm. I'm merging and releasing. Great job @bengro, thank you for your time spent on this.

**Comment by bengro** [comment]
👍  very cool, someone requests it, I'm happy to help!

Awesome! Thanks for your reviews and feedback. It definitely made it better. Can't wait to use it 😊

**Pull request merged**